src/module.c

```diff
@@ -3048,6 +3087,227 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
+ * - `flags`:
+ *   - `REDISMODULE_STREAM_EXCLUSIVE`: Don't include `start` and `end` in the
+ *     iterated range.
+ *   - `REDISMODULE_STREAM_REVERSE`: Iterate in reverse order, starting from
+ *     the `end` of the range.
+ * - `start`: The lower bound of the range. Use NULL for the beginning of the
+ *   stream.
+ * - `end`: The upper bound of the range. Use NULL for the end of the stream.
+ *
+ * Returns REDISMODULE_OK on success and REDISMODULE_ERR if the key doesn't
+ * refer to a stream or if invalid arguments were given.
+ *
+ * The stream items are retrieved using RedisModule_StreamIteratorNext(). The
+ * iterator must be freed by callling RedisModule_StreamIteratorStop().
+ */
+int RM_StreamIteratorStart(RedisModuleKey *key, int flags, RedisModuleStreamID *start, RedisModuleStreamID *end) {
+    /* check args */
+    if (flags & ~(REDISMODULE_STREAM_EXCLUSIVE |
+                  REDISMODULE_STREAM_REVERSE)) {
+        return REDISMODULE_ERR; /* invalid flags */
+    }
+    if (!key || !key->value || key->value->type != OBJ_STREAM) {
+        return REDISMODULE_ERR; /* not a stream */
+    }
+
+    /* define range for streamIteratorStart() */
+    streamID lower, upper;
+    if (start) lower = (streamID){start->ms, start->seq};
+    if (end)   upper = (streamID){end->ms,   end->seq};
+    if (flags & REDISMODULE_STREAM_EXCLUSIVE) {
+        if (start && streamIncrID(&lower) != C_OK) return REDISMODULE_ERR;
+        if (end   && streamDecrID(&upper) != C_OK) return REDISMODULE_ERR;
+    }
+
+    /* create iterator */
+    stream *s = key->value->ptr;
+    int rev = flags & REDISMODULE_STREAM_REVERSE;
+    streamIterator *si = zmalloc(sizeof(*si));
+    streamIteratorStart(si, s, start ? &lower : NULL, end ? &upper : NULL, rev);
+    key->iter = si;
+    return REDISMODULE_OK;
+}
+
+/* Stops a stream iterator created using RedisModule_StreamIteratorStart() and
+ * reclaims its memory.
+ *
+ * Returns REDISMODULE_OK on success and REDISMODULE_ERR if the key isn't
+ * refering to a stream or if no stream iterator has been started on the key. */
+int RM_StreamIteratorStop(RedisModuleKey *key) {
+    if (!key || !key->value || key->value->type != OBJ_STREAM || !key->iter) {
+        return REDISMODULE_ERR;
+    }
+    zfree(key->iter);
+    key->iter = NULL;
+    return REDISMODULE_OK;
+}
+
+/* Gets an item from a stream.
+ *
+ * - `key`: Key for which a stream iterator has been started using
+ *   RedisModule_StreamIteratorStart().
+ * - `maxnumfields`: The maximum number of fields to return in `fieldsvalues`.
+ * - `id`: The stream ID returned.
+ * - `fieldsvalues`: Pointer to an array where fields and values are returned in
+ *   the following order: (field 1, value 1, field 2, value 2, ...). The array
+ *   must have space for `2 * maxnumfields` elements. The strings returned are
+ *   freed automatically when the callback finishes if automatic memory
+ *   management has been enabled.
+ * - `numfields`: The number of fields returned.
```

Comment:
maybe a better API is to let the user interrogate the item at the iterator cursor rather than return it (or a portion of it) in the `next` call?
i.e. `RM_StreamIteratorNext` will return count, or just an ok / error, and you'll have another function to query the individual fields and values?
this way there's no need for a pre-allocated array, and the user can run on the result until he finds the records he's interested in (in case it's a specific one).

Reply:
Just like the internal streamIterator then. I like it. So how about:

```C
int RM_StreamIteratorNextID(RedisModuleKey *key, RedisModuleStreamID *id, long *numfields);
int RM_StreamIteratorNextField(RedisModuleKey *key, RedisModuleString *field, RedisModuleString *value);
```